src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/GerritTrigger.java

```diff
@@ -901,6 +908,92 @@ public boolean equals(Object obj) {
         return false;
     }
 
+    /**
+     * Get server by change.
+     *
+     * @param change the change
+     * @return Gerrit server.
+     */
+    private GerritServer getServerByChange(Change change) {
+        for (GerritServer server : PluginImpl.getServers_()) {
+            if (change.getUrl().contains(server.getHostName())) {
+                logger.trace("chnage: {}; server: {}", change, server.getHostName());
+                return server;
+            }
+        }
+        logger.error("can not find server for change: {}", change);
+        return null;
+    }
+
+    /**
+     * Should we trigger on this change?
+     *
+     * @param change the incoming change
+     * @param project the configured gerrit project
+     * @return true if we should.
+     */
+    private boolean isChangeInteresting(Change change, GerritProject project) {
+        boolean shouldTrigger = false;
+        boolean containsFilePathsOrForbiddenFilePaths = ((project.getFilePaths() != null
+                && project.getFilePaths().size() > 0)
+                || (project.getForbiddenFilePaths() != null && project.getForbiddenFilePaths().size() > 0));
+
+        if (isFileTriggerEnabled() && containsFilePathsOrForbiddenFilePaths) {
+            if (project.isInteresting(change.getProject(), change.getBranch(), change.getTopic(),
+                    change.getFiles(new GerritQueryHandler(getServerByChange(change).getConfig())))) {
+                shouldTrigger = true;
+            }
+        } else {
+            if (project.isInteresting(change.getProject(), change.getBranch(), change.getTopic())) {
+                shouldTrigger = true;
+            }
+        }
+
+        if (shouldTrigger) {
+            logger.trace("({},{},{}) is interesting; {}", change.getProject(),
+                    change.getBranch(), change.getTopic(), change);
+        }
+
+        return shouldTrigger;
+    }
+
+    /**
+     * Get related changes.
+     *
+     * @param changeBasedEvent the changeBasedEvent
+     * @return related changes.
+     */
+    private Set<Change> getChanges(ChangeBasedEvent changeBasedEvent) {
+        Set<Change> set = new HashSet<Change>();
+        Change change = changeBasedEvent.getChange();
+
+        set.add(change);
+
+        Topic topic = change.getTopicObject();
+        if (topic == null) {
+            logger.trace("{} has no topic", change);
+            return set;
+        }
+
+        for (GerritServer server : PluginImpl.getServers_()) {
+            topic = new Topic(topic.getName()); // Use a new topic object to avoid getting old changes
+            try {
+                logger.trace("query topic {} from {} ({}:{})", topic, server.getName(), server.getHostName(),
+                        server.getSshPort());
+                Map<Change, PatchSet> changes = topic.getChanges(new GerritQueryHandler(server.getConfig()));
```

Comment:
As @rsandell mentioned in https://github.com/jenkinsci/gerrit-trigger-plugin/pull/424:
This is not good. Think of a Jenkins controller with thousands of jobs listening to events. For every change based event thousands of jobs would be sending a query to Gerrit, sinking both it and Jenkins.
You need to do as is done with the file paths and only query once for each event and cache it (where the extra info is needed) so that the following jobs in the list won't have to make the query.

There is more work to be done.